// @prisma/instrumentation@5.14.0/dist/chunk-IEI5TMJX.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var chunk_IEI5TMJX_exports = {};
__export(chunk_IEI5TMJX_exports, {
  PrismaInstrumentation: () => PrismaInstrumentation
});
module.exports = __toCommonJS(chunk_IEI5TMJX_exports);
var import_chunk_VVAFFO6L = require("./chunk-VVAFFO6L.js");
var import_chunk_V6TNA4CL = require("./chunk-V6TNA4CL.js");
var import_instrumentation = require("@opentelemetry/instrumentation");
var PrismaInstrumentation = class extends import_instrumentation.InstrumentationBase {
  constructor(config = {}) {
    super(import_chunk_V6TNA4CL.NAME, import_chunk_V6TNA4CL.VERSION, config);
  }
  init() {
    const module2 = new import_instrumentation.InstrumentationNodeModuleDefinition(import_chunk_V6TNA4CL.MODULE_NAME, [import_chunk_V6TNA4CL.VERSION]);
    return [module2];
  }
  enable() {
    const config = this._config;
    const globalValue = {
      helper: new import_chunk_VVAFFO6L.ActiveTracingHelper({ traceMiddleware: config.middleware ?? false })
    };
    global[import_chunk_V6TNA4CL.GLOBAL_KEY] = globalValue;
  }
  disable() {
    delete global[import_chunk_V6TNA4CL.GLOBAL_KEY];
  }
  isEnabled() {
    return Boolean(global[import_chunk_V6TNA4CL.GLOBAL_KEY]);
  }
};
