// knip@5.17.4/dist/typescript/resolveModuleNames.js
import { existsSync } from 'node:fs';
import { isBuiltin } from 'node:module';
import ts from 'typescript';
import { DEFAULT_EXTENSIONS, FOREIGN_FILE_EXTENSIONS } from '../constants.js';
import { sanitizeSpecifier } from '../util/modules.js';
import { dirname, extname, isAbsolute, isInNodeModules, join } from '../util/path.js';
import { resolveSync } from '../util/resolve.js';
import { isDeclarationFileExtension } from './ast-helpers.js';
import { ensureRealFilePath, isVirtualFilePath } from './sys.js';
const resolutionCache = new Map();
const fileExists = (name, containingFile) => {
    const resolvedFileName = isAbsolute(name) ? name : join(dirname(containingFile), name);
    if (existsSync(resolvedFileName)) {
        return {
            resolvedFileName,
            extension: extname(name),
            isExternalLibraryImport: false,
            resolvedUsingTsExtension: false,
        };
    }
};
export function createCustomModuleResolver(customSys, compilerOptions, virtualFileExtensions, toSourceFilePath, useCache = true) {
    const extensions = [...DEFAULT_EXTENSIONS, ...virtualFileExtensions];
    function resolveModuleNames(moduleNames, containingFile) {
        return moduleNames.map(moduleName => {
            if (!useCache)
                return resolveModuleName(moduleName, containingFile);
            const key = moduleName.startsWith('.')
                ? join(dirname(containingFile), moduleName)
                : `${containingFile}:${moduleName}`;
            if (resolutionCache.has(key))
                return resolutionCache.get(key);
            const resolvedModule = resolveModuleName(moduleName, containingFile);
            if (resolvedModule)
                resolutionCache.set(key, resolvedModule);
            return resolvedModule;
        });
    }
    function resolveModuleName(name, containingFile) {
        const sanitizedSpecifier = sanitizeSpecifier(name);
        if (isBuiltin(sanitizedSpecifier) || isInNodeModules(name))
            return undefined;
        {
            const resolvedFileName = resolveSync(sanitizedSpecifier, containingFile, extensions);
            if (resolvedFileName) {
                const ext = extname(resolvedFileName);
                const extension = virtualFileExtensions.includes(ext) ? ts.Extension.Js : ext;
                if (!virtualFileExtensions.includes(ext) && !FOREIGN_FILE_EXTENSIONS.has(ext)) {
                    const srcFilePath = toSourceFilePath(resolvedFileName);
                    if (srcFilePath) {
                        return {
                            resolvedFileName: srcFilePath,
                            extension: extname(srcFilePath),
                            isExternalLibraryImport: false,
                            resolvedUsingTsExtension: false,
                        };
                    }
                }
                return {
                    resolvedFileName,
                    extension,
                    isExternalLibraryImport: isInNodeModules(resolvedFileName),
                    resolvedUsingTsExtension: false,
                };
            }
        }
        const tsResolvedModule = ts.resolveModuleName(sanitizedSpecifier, containingFile, compilerOptions, ts.sys).resolvedModule;
        if (tsResolvedModule) {
            if (isDeclarationFileExtension(tsResolvedModule.extension)) {
                const srcFilePath = toSourceFilePath(tsResolvedModule.resolvedFileName);
                if (srcFilePath) {
                    return {
                        resolvedFileName: srcFilePath,
                        extension: extname(srcFilePath),
                        isExternalLibraryImport: false,
                        resolvedUsingTsExtension: false,
                    };
                }
            }
            return tsResolvedModule;
        }
        const customResolvedModule = ts.resolveModuleName(sanitizedSpecifier, containingFile, compilerOptions, customSys).resolvedModule;
        if (!customResolvedModule || !isVirtualFilePath(customResolvedModule.resolvedFileName, virtualFileExtensions)) {
            const module = fileExists(sanitizedSpecifier, containingFile);
            if (module)
                return module;
            return customResolvedModule;
        }
        const resolvedFileName = ensureRealFilePath(customResolvedModule.resolvedFileName, virtualFileExtensions);
        const resolvedModule = {
            extension: ts.Extension.Js,
            resolvedFileName,
            isExternalLibraryImport: customResolvedModule.isExternalLibraryImport,
        };
        return resolvedModule;
    }
    return resolveModuleNames;
}
